src/test/scala/io/github/spark_redshift_community/spark/redshift/RedshiftSourceSuite.scala

```diff
@@ -0,0 +1,598 @@
+        " DISTSTYLE KEY DISTKEY \\(testint\\).*").r,
+      "COPY \"PUBLIC\"\\.\"test_table.*\"".r,
+      "GRANT SELECT ON \"PUBLIC\"\\.\"test_table\" TO jeremy".r)
+
+    val mockRedshift = new MockRedshift(
+      defaultParams("url"),
+      Map(TableName.parseFromEscaped("test_table").toString -> TestUtils.testSchema))
+
+    val relation = RedshiftRelation(
+      mockRedshift.jdbcWrapper,
+      _ => mockS3Client,
+      Parameters.mergeParameters(params),
+      userSchema = None)(testSqlContext)
+    relation.asInstanceOf[InsertableRelation].insert(expectedDataDF, overwrite = true)
+
+    // Make sure we wrote the data out ready for Redshift load, in the expected formats.
+    // The data should have been written to a random subdirectory of `tempdir`. Since we clear
+    // `tempdir` between every unit test, there should only be one directory here.
+    assert(s3FileSystem.listStatus(new Path(s3TempDir)).length === 1)
+    val dirWithAvroFiles = s3FileSystem.listStatus(new Path(s3TempDir)).head.getPath.toUri.toString
+    val written = testSqlContext.read.format("com.databricks.spark.avro").load(dirWithAvroFiles)
+    checkAnswer(written, TestUtils.expectedDataWithConvertedTimesAndDates)
+    mockRedshift.verifyThatConnectionsWereClosed()
+    mockRedshift.verifyThatExpectedQueriesWereIssued(expectedCommands)
+  }
+
+  test("Cannot write table with column names that become ambiguous under case insensitivity") {
+    val mockRedshift = new MockRedshift(
+      defaultParams("url"),
+      Map(TableName.parseFromEscaped("test_table").toString -> TestUtils.testSchema))
+
+    val schema = StructType(Seq(StructField("a", IntegerType), StructField("A", IntegerType)))
+    val df = testSqlContext.createDataFrame(sc.emptyRDD[Row], schema)
+    val writer = new RedshiftWriter(mockRedshift.jdbcWrapper, _ => mockS3Client)
+
+    intercept[IllegalArgumentException] {
+      writer.saveToRedshift(
+        testSqlContext, df, SaveMode.Append, Parameters.mergeParameters(defaultParams))
+    }
+    mockRedshift.verifyThatConnectionsWereClosed()
+    mockRedshift.verifyThatCommitWasNotCalled()
+    mockRedshift.verifyThatRollbackWasCalled()
+    mockRedshift.verifyThatExpectedQueriesWereIssued(Seq.empty)
+  }
+
+  test("Failed copies are handled gracefully when using a staging table") {
+    val params = defaultParams ++ Map("usestagingtable" -> "true")
+
+    val mockRedshift = new MockRedshift(
+      defaultParams("url"),
+      Map(TableName.parseFromEscaped("test_table").toString -> TestUtils.testSchema),
+      jdbcQueriesThatShouldFail = Seq("COPY \"PUBLIC\".\"test_table.*\"".r))
+
+    val expectedCommands = Seq(
+      "DROP TABLE IF EXISTS \"PUBLIC\".\"test_table.*\"".r,
+      "CREATE TABLE IF NOT EXISTS \"PUBLIC\".\"test_table.*\"".r,
+      "COPY \"PUBLIC\".\"test_table.*\"".r,
+      ".*FROM stl_load_errors.*".r
+    )
+//    testSqlContext = Mockito.mock(classOf[SQLContext])
```

Comment:
I don't think you need this line anymore

Reply:
good catch